File: worker/src/controllers/orderController.ts
import { ORDERBOOK } from "../models/orderbook";
import {
  buyNoOption,
  buyYesOption,
  sellNoOption,
  sellYesOption,
} from "../utils/helper";

export const buyOption = (userId:string, stockSymbol:string, quantity:number, originalPrice:number,stockType:string) => {

  const price = originalPrice / 100;
  let response;

  if (stockType == "yes") {
    response = buyYesOption(userId, stockSymbol, quantity, price);
  } else if (stockType == "no") {
    response = buyNoOption(userId, stockSymbol, quantity, price);
  }
  return {error:false, msg:response};
};

export const sellOption = (userId:string, stockSymbol:string, quantity:number, originalPrice:number,stockType:string) => {
  
  const price = originalPrice / 100;

  if (stockType == "yes") {
    const response = sellYesOption(userId, stockSymbol, quantity, price);
    return {error:false, msg: response};
  } else if (stockType == "no") {
    const response = sellNoOption(userId, stockSymbol, quantity, price)
    return {error:false, msg:response};
  }
};

export const viewOrderbook = () => {
  return ORDERBOOK;
};

export const viewIndividualOrderbook = (stockSymbol:string) => {
  const orderbook = ORDERBOOK[stockSymbol];

  if (!orderbook) {
    return { error:true, msg: "Orderbook with provided stock symbol not found" };
  }

  return {error:false, msg: orderbook};
};

File: go_worker/controllers/orderController.go
package controllers

import (
	"github.com/dipenbhat557/probo_clone/models"
	"github.com/dipenbhat557/probo_clone/utils"
)

func BuyOption(userId string, stockSymbol string, quantity int, originalPrice int, stockType string) map[string]interface{} {
	price := float64(originalPrice) / 100
	var response map[string]interface{}

	if stockType == "yes" {
		response = utils.BuyYesOption(userId, stockSymbol, quantity, price)
	} else if stockType == "no" {
		response = utils.BuyNoOption(userId, stockSymbol, quantity, price)
	}

	return map[string]interface{}{"error": false, "msg": response}
}

func SellOption(userId string, stockSymbol string, quantity int, originalPrice int, stockType string) map[string]interface{} {
	price := float64(originalPrice) / 100

	if stockType == "yes" {
		response := utils.SellYesOption(userId, stockSymbol, quantity, price)
		return map[string]interface{}{"error": false, "msg": response}
	} else if stockType == "no" {
		response := utils.SellNoOption(userId, stockSymbol, quantity, price)
		return map[string]interface{}{"error": false, "msg": response}
	}

	return map[string]interface{}{"error": true, "msg": "Invalid stock type"}
}

func ViewOrderbook() map[string]interface{} {
	res, err := utils.ConvertFullOrderBookToJSON()

	if err != nil {
		utils.HandleNilError(err)
	}

	return map[string]interface{}{"error": false, "msg": res}
}

func ViewIndividualOrderbook(stockSymbol string) map[string]interface{} {
	_, exists := models.ORDERBOOK[stockSymbol]

	if !exists {
		return map[string]interface{}{"error": true, "msg": "Orderbook with provided stock symbol not found"}
	}

	res, err := utils.ConvertIndividualOrderBookToJSON(stockSymbol)

	if err != nil {
		utils.HandleNilError(err)
	}

	return map[string]interface{}{"error": false, "msg": res}
}
